conceptnet_sme/semantic_matching.py

```diff
@@ -519,18 +537,322 @@ def export(self, dirname):
         save_hdf(related_terms, str(path / "terms-related.h5"))
 
 
-def train_model(model):
+
+def clip_grad_norm(parameters, max_norm, norm_type=2):
+    r"""Clips gradient norm of an iterable of parameters, just like 
+    nn.utils.clip_grad_norm_, but works even if some parameters are sparse.
+
+    The norm is computed over all gradients together, as if they were
+    concatenated into a single vector. Gradients are modified in-place.
+
+    Arguments:
+        parameters (Iterable[Tensor]): an iterable of Tensors that will have
+            gradients normalized
+        max_norm (float or int): max norm of the gradients
+        norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for
+            infinity norm.
+
+    Returns:
+        Total norm of the parameters (viewed as a single vector).
+    """
+    parameters = list(filter(lambda p: p.grad is not None, parameters))
+    max_norm = float(max_norm)
+    norm_type = float(norm_type)
+    if norm_type == float('inf'):
+        def L_inf_norm(tensor):
+            if tensor.layout == torch.sparse_coo:
+                if tensor._nnz() > 0:
+                    return tensor.coalesce()._values().abs().max()
+                else:
+                    return 0.0
+            else:
+                return tensor.abs().max()
+        total_norm = max(L_inf_norm(p.grad.data) for p in parameters)
+    else:
+        total_norm = 0
+        for p in parameters:
+            if p.grad.data.layout == torch.sparse_coo:
+                param_norm = p.grad.data.coalesce()._values().norm(norm_type)
+            else:
+                param_norm = p.grad.data.norm(norm_type)
+            total_norm += param_norm ** norm_type
+        total_norm = total_norm ** (1. / norm_type)
+    clip_coef = float(max_norm / (total_norm + 1e-6))
+    if clip_coef < 1:
+        for p in parameters:
+            p.grad.data.mul_(clip_coef)
+    return total_norm
+
+
+
+class SGD_Sparse(optim.SGD):
+    """
+    Just like optim.SGD except that this class handles non-zero 
+    weight decay and momentum even when some of the parameters 
+    being optimized have sparse gradient storage.
+    
+    NOTE:  Currently there seem to be gpu memory leaks when using 
+    Nesterov momentum; use that at your own risk.
+    """
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+
+            for p in group['params']:
+                if p.grad is None:
+                    continue
+                d_p = p.grad.data
```

Comment:
Is this code parallel to some existing code for SGD? Otherwise I find variable names such as `p` and `d_p` terribly opaque.

Reply:
Yes, I took the existing code for optim.SGD and tried to make a minimal set of changes and additions to support sparse gradients.